build/patches/add-user-scripts.v5.patch

```diff
@@ -0,0 +1,9810 @@
++const char kScriptIsEnabled[] = "enabled";
++const char kScriptName[] = "name";
++const char kScriptDescription[] = "description";
++const char kScriptVersion[] = "version";
++const char kScriptInstallTime[] = "install_time";
++
++class PrefUpdate : public DictionaryPrefUpdate {
++ public:
++  PrefUpdate(PrefService* service,
++             const std::string& id,
++             const std::string& path)
++      : DictionaryPrefUpdate(service, path), id_(id) {}
++
++  ~PrefUpdate() override = default;
++
++  base::DictionaryValue* Get() override {
++    base::DictionaryValue* dict = DictionaryPrefUpdate::Get();
++    base::Value* dict_item =
++        dict->FindKeyOfType(id_, base::Value::Type::DICTIONARY);
++    if (!dict_item)
++      dict_item = dict->SetKey(id_, base::Value(base::Value::Type::DICTIONARY));
++    return static_cast<base::DictionaryValue*>(dict_item);
++  }
++
++ private:
++  const std::string id_;
++
++  DISALLOW_COPY_AND_ASSIGN(PrefUpdate);
++};
++
++bool GetInt64FromPref(const base::DictionaryValue* dict,
++                      const std::string& key,
++                      int64_t* value) {
++  DCHECK(dict);
++  std::string value_str;
++  if (!dict->GetStringWithoutPathExpansion(key, &value_str)) {
++    VLOG(2) << "Can't find key in local pref dictionary. Invalid key: " << key
++            << ".";
++    return false;
++  }
++
++  if (!base::StringToInt64(value_str, value)) {
++    VLOG(2) << "Can't change string to int64_t. Invalid string value: "
++            << value_str << ".";
++    return false;
++  }
++
++  return true;
++}
++
++}
++
++UserScriptsPrefs::UserScriptsPrefs(
++    PrefService* prefs)
++    : prefs_(prefs) {
++}
++
++// static
++void UserScriptsPrefs::RegisterProfilePrefs(user_prefs::PrefRegistrySyncable* registry) {
++  registry->RegisterBooleanPref(kUserScriptsEnabled, false);
++  registry->RegisterDictionaryPref(kUserScriptsList);
++}
++
++bool UserScriptsPrefs::IsEnabled() {
++  return prefs_->GetBoolean(kUserScriptsEnabled);
++}
++
++void UserScriptsPrefs::SetEnabled(bool enabled) {
++  prefs_->SetBoolean(kUserScriptsEnabled, enabled);
++}
++
++void UserScriptsPrefs::CompareWithPrefs(UserScriptList& user_scripts) {
++  if( IsEnabled() == false ) {
++    LOG(INFO) << "UserScripts disabled";
++    user_scripts.clear();
++    return;
++  }
++
++  std::vector<std::string> all_scripts;
++
++  auto it = user_scripts.begin();
++  while (it != user_scripts.end())
++  {
++    std::string key = it->get()->key();
++    all_scripts.push_back(key);
++
++    std::unique_ptr<UserScriptsListPrefs::ScriptInfo> scriptInfo =
++        UserScriptsPrefs::CreateScriptInfoFromPrefs(key);
++
++    // add or update prefs
++    scriptInfo->set_name(it->get()->name());
++    scriptInfo->set_description(it->get()->description());
++    scriptInfo->set_version(it->get()->version());
++
++    // LOG(INFO) << "---UserScriptsPrefs::CompareWithPrefs name " << scriptInfo->name();
```

Comment:
Please uncomment or delete this line.

Reply:
fixed in https://github.com/bromite/bromite/commit/2812a0f9a76a7b3052b7a408344095680e7244ff